### PR TITLE
Set up CTest

### PR DESCRIPTION
**Pull request by EZoni**
Work-in-progress PR to set up CTest. Apply the same logic as in ImpactX and adjust/advance for WarpX.

- Example of current input files organization:
  - `inputs_base_2d`
    (base input file for shared input parameters named as `inputs_base_<1d;2d;3d;rz>`)
  - `inputs_test_2d_langmuir_multi_nodal`,
    `inputs_test_2d_langmuir_multi_psatd`,
    ...
    (test input files for test-specific input parameters named as `inputs_test_<1d;2d;3d;rz>_...`, read base input file through the `FILE` parameter)
  - test names must start with `test_<1d;2d;3d;rz>_...` to guarantee consistency and make maintenance easier

- New Azure CI matrix elements (list coming soon).

- Run all MPI tests with 2 MPI processes and 1 OpenMP thread. Run single-process tests without MPI.

To-do:
- [x] `WarpX_TEST_CLEANUP` to remove test directories
- [x] `WarpX_TEST_FPETRAP` for runtime fpe-trapping
- [x] Resolve `FIXME` comments or redirect to new issues
- [x] Avoid copying inputs files https://github.com/AMReX-Codes/amrex/pull/4126
- [x] Enforce new naming convention through CI style check
- [x] Remove obsolete CI style check scripts
- [x] Remove obsolete regression testing scripts: #5220
- [x] Update docs
- [x] Remove files after each test: new build flag `WarpX_CI`
- [x] Make sure we don't miss any build variant:
  - [x] single-process Python
  - [x] embedded boundaries
  - single- & mixed-precision (new issue #5205)
- [x] Add missing checksums (new issue #5206):
  - `dirichlet_bc`
  - `dirichlet_bc_picmi`
  - `field_probe`
  - `embedded_boundary_picmi`
  - `particle_absorption`
  - `particle_reflection_picmi`
  - [x] `id_cpu_read_picmi`
  - `runtime_components_picmi`
  - `laser_acceleration_rz_opmd`
- [x] Support restart tests
  - [x] Separate input file based on original input file (read through the `FILE` parameter)
  - [x] Call default regression analysis against original checksum file, with small tolerance
  - [x] Support restart from PICMI (can we set `FILE` through PICMI?)
  - [x] Update `analysis_default_restart.py`
- [x] Support custom run commands (e.g., custom PICMI scripts using lists vs strings)
- Add stricter dependency between `.analysis` test and `.run` test:
  run `.analysis` test only if `.run` test passed (now runs if `.run` test ran)
- Update test README files (e.g., Ohm solver tests)
- [ ] Skip pyAMReX tests directly from CMakeLists.txt

### CODE REVIEW

File: Examples/CMakeLists.txt
Comment (ax3l) [inline Examples/CMakeLists.txt]: I would add this if inside the function that is called. The user passed the dim, if it does not match the list of compiled dims return (see: MPI).

File: Examples/CMakeLists.txt
Comment (ax3l) [inline Examples/CMakeLists.txt]: As in ImpactX, I would not add separate variables before the function call to keep these entries per test compact. Add inline / end of line comments where needed.

File: Regression/WarpX-tests.ini
Comment (EZoni) [inline Regression/WarpX-tests.ini]: @ax3l @RemiLehe 

By the way, I think this type of input files organization (which came up here just as a way to guarantee backwards compatibility while setting up CTest) is quite good regardless of CTest itself.

At least all input parameters can be found in files from the same directory (e.g., Examples/Tests/langmuir/), rather than combining files from two directories (e.g., Examples/Tests/langmuir/ and Regression/). This should make it easier to reproduce a test locally.

It makes it also easier to spot differences between two tests that share the same base input deck (do a `diff` of the two test-specific additional files, rather than compare the `runtime_params` lines of the two tests in the INI file).

I just wanted to point this out so that it doesn't get lost within the other (main) changes in this PR.

Comment (ax3l) [reply Regression/WarpX-tests.ini]: Discussed in person:

We will use `FILE=` in the concrete input file we register with a test to reference a base parameter set (from another file). That way, we can avoid all CLI extra parameters and extra options.

File: .azure-pipelines.yml
Comment (EZoni) [inline .azure-pipelines.yml:133]: Clear pip cache (see #5110):
```suggestion
      python3 -m pip install --upgrade -r Regression/requirements.txt
      python3 -m pip cache purge
```

File: Examples/Tests/nuclear_fusion/analysis_proton_boron_fusion.py
Comment (EZoni) [reply Examples/Tests/nuclear_fusion/analysis_proton_boron_fusion.py]: Fixed in [7903470](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/7903470d72e12b0e1690cf174bfd4abf418f9d92).

File: Examples/Tests/nuclear_fusion/analysis_proton_boron_fusion.py
Comment (EZoni) [reply Examples/Tests/nuclear_fusion/analysis_proton_boron_fusion.py]: Fixed in [ddacf8d](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/ddacf8d01c0d0d4ce62980dfc337b2c9ede93151).

File: .github/workflows/ubuntu.yml
Comment (EZoni) [inline .github/workflows/ubuntu.yml]: I had to ~~comment this out~~ remove this because we don't have a guard that checks whether `WarpX_APP` is defined when we add the no-Python tests. We might want to add a check for that in case someone compiles with `-DWarpX_APP=OFF` locally. If that is the case, the configuration currently fails because it tries to find app executables that do not exist.

File: Examples/Tests/ion_stopping/analysis.py
Comment (ax3l) [reply Examples/Tests/ion_stopping/analysis.py:153]:
```suggestion
for _ in range(last_it):
```

File: Examples/CMakeLists.txt
Comment (ax3l) [inline Examples/CMakeLists.txt]: Discussed today: eventually, we want to remove this convention and just depend on `base` inputs and dependent tests (`dependency`).

Likely a follow-up PR to avoid changing too many tests with this one.

Comment (EZoni) [reply Examples/CMakeLists.txt]: Just a note to keep the conversation going. 

We can discuss this, but I do think that enforcing the convention (currently already implemented in the new automated style checks - which, by the way, I rewrote as a single Python script `check_inputs.py` in this PR) has advantages.

Right now one can type
```
ctest --test-dir build/ --output-on-failure -R test_3d_acceleration_psatd
```
and get this output, which I think is pretty easy to parse:
```
Internal ctest changing into directory: /home/edoardo/src/warpx/build
Test project /home/edoardo/src/warpx/build
      Start 506: test_3d_acceleration_psatd.run
 1/12 Test #506: test_3d_acceleration_psatd.run .........................   Passed    3.69 sec
      Start 507: test_3d_acceleration_psatd.analysis
 2/12 Test #507: test_3d_acceleration_psatd.analysis ....................   Passed    2.27 sec
      Start 509: test_3d_acceleration_psatd_restart.run
 3/12 Test #509: test_3d_acceleration_psatd_restart.run .................   Passed    2.39 sec
      Start 510: test_3d_acceleration_psatd_restart.analysis
 4/12 Test #510: test_3d_acceleration_psatd_restart.analysis ............   Passed    2.32 sec
      Start 511: test_3d_acceleration_psatd_restart.cleanup
 5/12 Test #511: test_3d_acceleration_psatd_restart.cleanup .............   Passed    0.06 sec
      Start 508: test_3d_acceleration_psatd.cleanup
 6/12 Test #508: test_3d_acceleration_psatd.cleanup .....................   Passed    0.07 sec
      Start 512: test_3d_acceleration_psatd_time_avg.run
 7/12 Test #512: test_3d_acceleration_psatd_time_avg.run ................   Passed    4.88 sec
      Start 513: test_3d_acceleration_psatd_time_avg.analysis
 8/12 Test #513: test_3d_acceleration_psatd_time_avg.analysis ...........   Passed    2.25 sec
      Start 515: test_3d_acceleration_psatd_time_avg_restart.run
 9/12 Test #515: test_3d_acceleration_psatd_time_avg_restart.run ........   Passed    2.94 sec
      Start 516: test_3d_acceleration_psatd_time_avg_restart.analysis
10/12 Test #516: test_3d_acceleration_psatd_time_avg_restart.analysis ...   Passed    2.24 sec
      Start 517: test_3d_acceleration_psatd_time_avg_restart.cleanup
11/12 Test #517: test_3d_acceleration_psatd_time_avg_restart.cleanup ....   Passed    0.06 sec
      Start 514: test_3d_acceleration_psatd_time_avg.cleanup
12/12 Test #514: test_3d_acceleration_psatd_time_avg.cleanup ............   Passed    0.11 sec

100% tests passed, 0 tests failed out of 12
```

If we allow arbitrary names for the restart test based on a given test (e.g., with a "restart" keyword anywhere in the test name - `test_3d_restart_acceleration_psatd` instead of `test_3d_acceleration_psatd_restart`), I think this output will be less easy to parse and/or work with (when it comes to parsing visually, parsing through code, sorting alphabetically, filtering tests in CTest via `-R`, etc.).

The convention does not seem like a big burden to me, the automated style checks can catch instances where it is not followed and reviewers/maintainers can make sure that such instances are fixed - it doesn't have to be an extra task on the developer's end.

File: Examples/CMakeLists.txt
Comment (ax3l) [inline Examples/CMakeLists.txt]: Does that help? (Also: this is a very low risk here.)
```suggestion
    if(python)
        set(THIS_Python_EXE ${Python_EXECUTABLE})
    else()
        set(THIS_Python_EXE "")
    endif()
```

File: Examples/CMakeLists.txt
Comment (ax3l) [inline Examples/CMakeLists.txt]: As above?
```suggestion
    if(python OR WIN32)
        set(THIS_Python_SCRIPT_EXE ${Python_EXECUTABLE})
    else()
        set(THIS_Python_SCRIPT_EXE "")
    endif()
```

**Comment by ax3l** [comment]
Aha, a lot of tests are too unstable to survive a change from Ubuntu 20.04 with GCC 9 to Ubuntu 22.04 with GCC 12. Will update Azure pipelines for speedy dependency install in a follow-up PR #5203. Documented failing tests for developer meeting and for hackathon.

File: Examples/CMakeLists.txt
Comment (ax3l) [inline Examples/CMakeLists.txt:149]: I currently have no great idea how to implement this without major hackery.

File: Examples/CMakeLists.txt
Comment (ax3l) [inline Examples/CMakeLists.txt]: The first `${PYTHONPATH}` will be undefined unless we read it first. Do you mean
```suggestion
        set_property(TEST ${name}.run APPEND PROPERTY ENVIRONMENT "PYTHONPATH=$ENV{PYTHONPATH}:${CMAKE_PYTHON_OUTPUT_DIRECTORY}")
```

File: Examples/CMakeLists.txt
Comment (EZoni) [reply Examples/CMakeLists.txt:149]: Maybe I can open another follow-up issue and we can think about how to put this back in after the first merge.

Comment (ax3l) [reply Examples/CMakeLists.txt:149]: Since we only need this for CI:
An idea could be to add default-value control via an env variable for this option. Then overwrite it with inputs.

File: Docs/source/developers/testing.rst
Comment (ax3l) [inline Docs/source/developers/testing.rst]: CMake is a supported syntax :)
```suggestion
  .. code-block:: cmake
```

File: Docs/source/developers/testing.rst
Comment (ax3l) [inline Docs/source/developers/testing.rst]: I think these are overindented blocks.

File: Docs/source/developers/testing.rst
Comment (ax3l) [inline Docs/source/developers/testing.rst]: Similar to Python syntax and for readability, I would do a double space before `#` inline comments (here and in all tests).
```suggestion
           test_1d_laser_acceleration  # name
           1  # dims
           2  # nprocs
           OFF  # eb
           Examples/Tests/laser_acceleration/inputs_test_1d_laser_acceleration  # inputs
           Examples/Tests/laser_acceleration/analysis.py  # analysis
           diags/diag1000100  # output (plotfile)
           OFF  # dependency
```

File: Docs/source/developers/testing.rst
Comment (ax3l) [inline Docs/source/developers/testing.rst]:
```suggestion
  .. code-block:: cmake
```

File: Docs/source/developers/testing.rst
Comment (ax3l) [inline Docs/source/developers/testing.rst]:
```suggestion
           test_2d_laser_acceleration_picmi  # name
           2  # dims
           2  # nprocs
           OFF  # eb
           Examples/Tests/laser_acceleration/inputs_test_2d_laser_acceleration_picmi.py  # inputs
           Examples/Tests/laser_acceleration/analysis.py  # analysis
           diags/diag1000100  # output (plotfile)
           OFF  # dependency
```

File: Docs/source/developers/testing.rst
Comment (ax3l) [inline Docs/source/developers/testing.rst]:
```suggestion
  .. code-block:: cmake
```

File: Docs/source/developers/testing.rst
Comment (ax3l) [inline Docs/source/developers/testing.rst]:
```suggestion
           test_3d_laser_acceleration_restart  # name
           3  # dims
           2  # nprocs
           OFF  # eb
           Examples/Tests/laser_acceleration/inputs_test_3d_laser_acceleration_restart  # inputs
           Examples/analysis_default_restart.py  # analysis
           diags/diag1000100  # output (plotfile)
           test_3d_laser_acceleration  # dependency
```

File: Docs/source/developers/testing.rst
Comment (ax3l) [inline Docs/source/developers/testing.rst]:
```suggestion
  .. code-block:: cmake
```

File: Docs/source/developers/testing.rst
Comment (ax3l) [inline Docs/source/developers/testing.rst]:
```suggestion
           test_rz_laser_acceleration_picmi  # name
           RZ  # dims
           2   # nprocs
           OFF  # eb
           "Examples/Tests/laser_acceleration/inputs_test_rz_laser_acceleration_picmi.py --test --dir 1"  # inputs
           Examples/Tests/laser_acceleration/analysis.py  # analysis
           diags/diag1/  # output (openPMD time series)
           OFF  # dependency
```

File: Docs/source/developers/testing.rst
Comment (ax3l) [inline Docs/source/developers/testing.rst]:
```suggestion
Note that we currently obey the following snake\_case naming conventions for test names and test input files (which make automation tasks easier, e.g., parsing visually, parsing through code, sorting alphabetically, filtering tests in CTest via ``-R``, etc.):
```

File: Examples/CMakeLists.txt
Comment (EZoni) [reply Examples/CMakeLists.txt]: Thanks, corrected in [ffd4d03](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/ffd4d031c42002c30008115d96921116ae3b3da7).

File: Docs/source/developers/testing.rst
Comment (EZoni) [reply Docs/source/developers/testing.rst]: Done in [8e64870](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/8e64870fe4fe42fd6c88f69741fe7761d30226c6).

File: Docs/source/developers/testing.rst
Comment (EZoni) [reply Docs/source/developers/testing.rst]: Done in [8e64870](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/8e64870fe4fe42fd6c88f69741fe7761d30226c6).

File: Docs/source/developers/testing.rst
Comment (EZoni) [reply Docs/source/developers/testing.rst]: Done in [8e64870](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/8e64870fe4fe42fd6c88f69741fe7761d30226c6).

File: Docs/source/developers/testing.rst
Comment (EZoni) [reply Docs/source/developers/testing.rst]: Done in [8e64870](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/8e64870fe4fe42fd6c88f69741fe7761d30226c6).

File: Docs/source/developers/testing.rst
Comment (EZoni) [reply Docs/source/developers/testing.rst]: Done in [8e64870](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/8e64870fe4fe42fd6c88f69741fe7761d30226c6).

File: Docs/source/developers/testing.rst
Comment (EZoni) [reply Docs/source/developers/testing.rst]: Done in [8e64870](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/8e64870fe4fe42fd6c88f69741fe7761d30226c6).

File: Docs/source/developers/testing.rst
Comment (EZoni) [reply Docs/source/developers/testing.rst]: Done in [8e64870](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/8e64870fe4fe42fd6c88f69741fe7761d30226c6).

File: Docs/source/developers/testing.rst
Comment (EZoni) [reply Docs/source/developers/testing.rst]: Done in [8e64870](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/8e64870fe4fe42fd6c88f69741fe7761d30226c6).

File: Docs/source/developers/testing.rst
Comment (EZoni) [reply Docs/source/developers/testing.rst]: Done in [8e64870](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/8e64870fe4fe42fd6c88f69741fe7761d30226c6).

File: Examples/CMakeLists.txt
Comment (EZoni) [reply Examples/CMakeLists.txt:149]: New issue: #5218

File: CMakeLists.txt
Comment (ax3l) [inline CMakeLists.txt:798]: Do we have this somewhere?
```suggestion
include(CTest)
if(BUILD_TESTING)
```
https://cmake.org/cmake/help/latest/module/CTest.html

Comment (ax3l) [reply CMakeLists.txt:798]: Ah, can you comment in line 199 please?

This is currently probably "luckily" activated due to the superbuild from another project, but we should add this ourselves.

Comment (EZoni) [reply CMakeLists.txt:798]: Done in [d63c920](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/d63c920d81e25877bfa54d7fa8d36ee6a9379497).

File: Docs/source/developers/testing.rst
Comment (EZoni) [reply Docs/source/developers/testing.rst]: Done in [cb5e42c](https://github.com/ECP-WarpX/WarpX/pull/5068/commits/cb5e42cf8e2102a3040dea7032404911fe17ef5b).

**Review by ax3l — APPROVED** [review]
Thanks a lot, fantastic work! :rocket: :sparkles:

**Pull request merged**
